File: packages/oxlint-plugin-react-doctor/src/plugin/rules/security-scan/public-env-secret-name.ts
import { defineRule } from "../../utils/define-rule.js";
import { findSuspiciousPublicEnvSecretNamePattern } from "./utils/find-suspicious-public-env-secret-name.js";
import { getMatchLocation } from "./utils/get-match-location.js";
import { isClientSourcePath } from "./utils/is-client-source-path.js";

// Source files under a docs tree are onboarding snippets, not shipped code.
const DOCS_DIRECTORY_PATTERN = /(?:^|\/)docs?\//i;

export const publicEnvSecretName = defineRule({
  id: "public-env-secret-name",
  title: "Secret-like public env variable",
  severity: "warn",
  recommendation:
    "Public env prefixes are inlined into browser bundles. Rename public values to non-secret names, and keep tokens, passwords, private keys, and service-role credentials server-only.",
  scan: (file) => {
    if (!isClientSourcePath(file.relativePath)) return [];
    if (DOCS_DIRECTORY_PATTERN.test(file.relativePath)) return [];
    const pattern = findSuspiciousPublicEnvSecretNamePattern(file.content);
    if (pattern === undefined) return [];

    const location = getMatchLocation(file.content, pattern);
    return [
      {
        message:
          "Client code references a public env variable whose name looks like a secret or privileged credential.",
        line: location.line,
        column: location.column,
      },
    ];
  },
});

File: native/oxlint/scans/public-env-secret-name.rs
use super::{ScanFinding, get_location_at_index::get_location_at_index};

const MESSAGE: &str = "Client code references a public env variable whose name looks like a secret or privileged credential.";

pub fn scan(relative_path: &str, source: &str) -> Vec<ScanFinding> {
    let normalized_path =
        super::normalize_js_regex_content::normalize_js_regex_content(relative_path);
    if !super::is_client_source_path::is_client_source_path(relative_path)
        || lazy_regex::regex_is_match!(r"(?i)(?:^|/)docs?/", normalized_path.as_ref())
    {
        return Vec::new();
    }
    let Some(index) =
        super::security_secret_patterns::first_suspicious_public_env_secret_name(source)
    else {
        return Vec::new();
    };
    let (line, column) = get_location_at_index(source, source, index);
    vec![ScanFinding::inherited(MESSAGE, line, column)]
}
